h5py/h5t.pyx

```diff
@@ -1106,7 +1106,17 @@ cdef class TypeCompoundID(TypeCompositeID):
         else:
             if sys.version[0] == '3':
                 field_names = [x.decode('utf8') for x in field_names]
-            typeobj = dtype({'names': field_names, 'formats': field_types, 'offsets': field_offsets})
+            if len(field_names) > 0:
+                collated_fields = zip(field_names, field_types, field_offsets)
+                ordered_fields = sorted(
+                    collated_fields, key=operator.itemgetter(2))
+                field_names, field_types, field_offsets = \
+                    [list(t) for t in zip(*ordered_fields)]
```

Comment:
That won't work, but `map(list, zip(*ordered_fields))` will. I'll replace it with the more concise version...